### PR TITLE
[XRay][X86] Fix stack alignment for custom event calls.

### PR DESCRIPTION
**Pull request by rickyz**
Calls to @llvm.xray.{custom,typed}event are lowered to an nop sled that can be
patched with a call to an instrumentation function at runtime. Prior to this
change, x86 codegen did not guarantee that these patched calls run with the
appropriate stack alignment (particularly in leaf functions, where normal stack
alignment assumptions may not hold).

This leads to crashes on x86, as the custom event hook can end up using SSE
instructions that assume a 16-byte aligned stack.

Fix this by wrapping custom event hooks in CALLSEQ_START/CALLSEQ_END as done for
regular function calls. One downside of this approach is that on functions whose
stacks aren't already aligned, we may end up running stack alignment fixup
instructions even when instrumentation is disabled. An alternative could be to
make the custom event assembly trampolines stack-alignment-agnostic. This was
the case in the past, but
https://github.com/llvm/llvm-project/commit/b46c89892fe25bec197fd30f09b3a312da126422
removed this due to complexity in maintaining CFI directives for these stack
adjustments. Since we are already willing to pay the call argument setup cost
for custom hooks when instrumentation is disabled, I am hoping that an extra
push/pop in this hopefully uncommon unaligned stack case is tolerable.

### CODE REVIEW

**Comment by rickyz** [comment]
@llvm/pr-subscribers-xray since the paths at https://github.com/orgs/llvm/teams/pr-subscribers-xray don't seem to match this PR.

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 1]
```diff
@@ -36254,6 +36254,31 @@ X86TargetLowering::EmitSjLjDispatchBlock(MachineInstr &MI,
...
+X86TargetLowering::emitPatchableEventCall(MachineInstr &MI,
+                                          MachineBasicBlock *BB) const {
+  // Wrap patchable event calls in CALLSEQ_START/CALLSEQ_END, as tracing
+  // calls may require proper stack alignment.
+  const TargetInstrInfo &TII = *Subtarget.getInstrInfo();
+  const MIMetadata MIMD(MI);
+  MachineFunction &MF = *BB->getParent();
+
+  // Emit CALLSEQ_START right before the instruction.
+  BB->getParent()->getFrameInfo().setAdjustsStack(true);
```
Comment (MaskRay) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:36267]: We can use here `MF.getFrameInfo()` here

File: llvm/test/CodeGen/X86/xray-leaf-stack-alignment.ll [hunk 2]
```diff
@@ -0,0 +1,18 @@
+; RUN: llc -mtriple=x86_64 < %s | FileCheck %s
```
Comment (MaskRay) [inline llvm/test/CodeGen/X86/xray-leaf-stack-alignment.ll:1]: A file-level comment `;; ...` can be useful

**Review by MaskRay — APPROVED** [review]
Thanks! This is similar to EmitLoweredTLSAddr

File: llvm/test/CodeGen/X86/xray-leaf-stack-alignment.ll (hunk 2 above)
Comment (MaskRay) [reply llvm/test/CodeGen/X86/xray-leaf-stack-alignment.ll:1]: The test can be added to xray-custom-log.ll

**Comment by rickyz** [comment]
> Thanks! This is similar to EmitLoweredTLSAddr

Yeah, I copied the approach taken there - I kept them separate mostly because I couldn't think of a great name capturing a nice conceptual overlap ("emit thing that will turn into a function call, either at compile time or in the case of xray, conditionally at runtime") - I guess I can also imagine the two diverging in the future - but lmk if you prefer otherwise!

File: llvm/test/CodeGen/X86/xray-leaf-stack-alignment.ll (hunk 2 above)
Comment (MaskRay) [reply llvm/test/CodeGen/X86/xray-leaf-stack-alignment.ll:1]: Move the test to xray-custom-log.ll ? :)

Comment (MaskRay) [reply llvm/test/CodeGen/X86/xray-leaf-stack-alignment.ll:1]: Per discussions on https://discourse.llvm.org/t/rfc-github-pr-resolve-conversation-button/73178 , it's recommended for reviewers to resolve comments, even if they are really straightforward:)

Comment (rickyz) [reply llvm/test/CodeGen/X86/xray-leaf-stack-alignment.ll]: Ah right, moved, thanks (and thanks for the heads up re resolving comments - still getting used to using Github :-))

**Review by MaskRay — APPROVED** [review]

**Comment by MaskRay** [comment]
> > Thanks! This is similar to EmitLoweredTLSAddr
> 
> Yeah, I copied the approach taken there - I kept them separate mostly because I couldn't think of a great name capturing a nice conceptual overlap ("emit thing that will turn into a function call, either at compile time or in the case of xray, conditionally at runtime") - I guess I can also imagine the two diverging in the future - but lmk if you prefer otherwise!

Adding a new function seems fine for now!

**Pull request merged**